presto-main/src/main/java/com/facebook/presto/sql/planner/CanonicalPlanGenerator.java

```diff
@@ -88,6 +95,61 @@ public Optional<PlanNode> visitPlan(PlanNode node, Context context)
         return Optional.empty();
     }
 
+    @Override
+    public Optional<PlanNode> visitUnion(UnionNode node, Context context)
+    {
+        if (strategy == DEFAULT) {
+            return Optional.empty();
+        }
+
+        // We want to order sources in a consistent manner. Variable names and plan node ids can mess with that because of
+        // our stateful canonicalization using `variableAllocator` and `planNodeIdAllocator`.
+        // So, we first try to canonicalize each source independently. They may have conflicting variable names, but we only use it
+        // to decide order, and then canonicalize them properly again.
+        // This can lead to O(n * h) time complexity, where n is number of plan nodes, and h is height of plan tree. This
+        // is at par with other pieces like hashing each plan node.
+        Multimap<String, Integer> sourceToPosition = TreeMultimap.create();
+        for (int i = 0; i < node.getSources().size(); ++i) {
+            PlanNode source = node.getSources().get(i);
+            Optional<CanonicalPlan> canonicalSource = generateCanonicalPlan(source, strategy, objectMapper);
+            if (!canonicalSource.isPresent()) {
+                return Optional.empty();
+            }
+            sourceToPosition.put(canonicalSource.get().toString(objectMapper), i);
+        }
```

Comment:
Can we
1. use the visitor pattern to `accept` all sources
2. wrap each source into `CanonicalPlan`
3. generate the hash and order of them
4. then sort?

This can avoid canonicalizing the sources twice right?

Reply:
Its because of variable names, which are like `expr_...`. It seems hard to do without some kind of variable re-mapping.